libusb/hid.c

```diff
@@ -912,6 +1086,346 @@ void  HID_API_EXPORT hid_free_enumeration(struct hid_device_info *devs)
+	} else {
+		hid_hotplug_context.queue = msg;
+	}
+	hidapi_thread_mutex_unlock(&hid_hotplug_context.libusb_thread);
+
+	/* Wake up the other thread so it can react to the new message immediately */
+	hidapi_thread_cond_signal(&hid_hotplug_context.callback_thread);
+
+	return 0;
+}
+
+static void process_hotplug_event(struct hid_hotplug_queue* msg)
+{
+	if (msg->event == LIBUSB_HOTPLUG_EVENT_DEVICE_ARRIVED) {
+		struct hid_device_info* info = hid_enumerate_from_libusb(msg->device, 0, 0);
+		struct hid_device_info* info_cur = info;
+		while (info_cur) {
+			/* For each device, call all matching callbacks */
+			/* TODO: possibly make the `next` field NULL to match the behavior on other systems */
+			hid_internal_invoke_callbacks(info_cur, HID_API_HOTPLUG_EVENT_DEVICE_ARRIVED);
+			info_cur = info_cur->next;
+		}
+
+		/* Append all we got to the end of the device list */
+		if (info) {
+			if (hid_hotplug_context.devs != NULL) {
+				struct hid_device_info* last = hid_hotplug_context.devs;
+				while (last->next != NULL) {
+					last = last->next;
+				}
+				last->next = info;
+			}
+			else {
+				hid_hotplug_context.devs = info;
+			}
+		}
+	}
+	else if (msg->event == LIBUSB_HOTPLUG_EVENT_DEVICE_LEFT) {
+		for (struct hid_device_info **current = &hid_hotplug_context.devs; *current;) {
+			struct hid_device_info* info = *current;
+			if (match_libusb_to_info(msg->device, *current)) {
+				/* If the libusb device that's left matches this HID device, we detach it from the list */
+				*current = (*current)->next;
+				info->next = NULL;
+				hid_internal_invoke_callbacks(info, HID_API_HOTPLUG_EVENT_DEVICE_LEFT);
+				/* Free every removed device */
+				free(info);
+			} else {
+				current = &info->next;
+			}
+		}
+	}
+
+	/* Release the libusb device - we are done with it */
+	libusb_unref_device(msg->device);
+
+	/* Cleanup note: this function is called inside a thread that the clenup function would be waiting to finish */
+	/* Any callbacks that await removal are removed in hid_internal_invoke_callbacks */
+	/* No further cleaning is needed */
+}
+
+static void* callback_thread(void* user_data)
+{
+	(void) user_data;
+
+	/* 5 msec timeout seems reasonable; don't set too low to avoid high CPU usage */
+	/* This timeout only affects how much time it takes to stop the thread */
+	hidapi_timespec ts;
+	ts.tv_sec = 0;
+	ts.tv_nsec = 5000000;
+
+	hidapi_thread_mutex_lock(&hid_hotplug_context.callback_thread);
+	
+	/* We stop the thread if by the moment there are no events left in the queue there are no callbacks left */
+	while (1) {
+		/* Make the tread fall asleep and wait for a condition to wake it up */
+		hidapi_thread_cond_timedwait(&hid_hotplug_context.callback_thread, &ts);
+
+		/* We use this thread's mutex to protect the queue */
+		hidapi_thread_mutex_lock(&hid_hotplug_context.libusb_thread);
+		while (hid_hotplug_context.queue) {
+			struct hid_hotplug_queue *cur_event = hid_hotplug_context.queue;
+			process_hotplug_event(cur_event);
```

Comment:
would it be possible/feasible to postpone the processing of the event until libusb has closed the device? I know that might get complicate but otherwise enumerate would not work (at least as long as https://github.com/libusb/libusb/pull/1532 has not been merged).

Reply:
I don't follow your thoughts here. Can you elaborate more details?

Reply:
well, libusb opens the device and fires the callback, then hidapi tries to get infos from the device but may fail as it is still open by libusb.
if I run my example program with the libusb backend, I don't get e.g. manufacturer infos but with the hidraw backend I get them.

Reply:
I was wrong, libusb does not open the device. but for some reason, `hid_enumerate_from_libusb()` is not able to open the device in my case (LIBUSB_ERROR_ACCESS) when called by the hotplug handler. If I insert a small delay before `libusb_open()` it works.

Reply:
Interesting finding...
As per [documentation](https://libusb.sourceforge.io/api-1.0/libusb_hotplug.html) - it should be able to use libusb_open from a hotplug callback function.
What version of libusb are you using?

Reply:
I use libusb v1.0.27 and this branch for hidapi. Everything can be found here: https://github.com/bearsh/hid/tree/feature/connection-callback including the test program. The project is a go-wrapper around hiadpi. The test application is in `cmd/hid-hotplug`.

Reply:
something like the following would solve my issue and is also suggested as workaround by @mcuee in https://github.com/libusb/libusb/issues/1586#issuecomment-2500248425

```diff
diff --git a/hidapi/libusb/hid.c b/hidapi/libusb/hid.c
index ca8555c..b1c31c5 100644
--- a/hidapi/libusb/hid.c
+++ b/hidapi/libusb/hid.c
@@ -944,7 +944,7 @@ static int should_enumerate_interface(unsigned short vendor_id, const struct lib
 	return 0;
 }
 
-static struct hid_device_info* hid_enumerate_from_libusb(libusb_device *dev, unsigned short vendor_id, unsigned short product_id)
+static struct hid_device_info* hid_enumerate_from_libusb(libusb_device *dev, unsigned short vendor_id, unsigned short product_id, int hotplug)
 {
 	struct hid_device_info *root = NULL; /* return object */
 	struct hid_device_info *cur_dev = NULL;
@@ -977,7 +977,11 @@ static struct hid_device_info* hid_enumerate_from_libusb(libusb_device *dev, uns
 				if (should_enumerate_interface(dev_vid, intf_desc)) {
 					struct hid_device_info *tmp;
 
-					res = libusb_open(dev, &handle);
+					/* after a hotplug event, retry it 5 time (max 50ms extra latency) */
+					unsigned try = hotplug ? 6 : 1;
+					while (try-- && (res = libusb_open(dev, &handle)) == LIBUSB_ERROR_ACCESS) {
+						usleep(10000);
+					}
 #ifdef __ANDROID__
 					if (handle) {
 						/* There is (a potential) libusb Android backend, in which
@@ -1058,7 +1062,7 @@ struct hid_device_info HID_API_EXPORT *hid_enumerate(unsigned short vendor_id, u
 		return NULL;
 
 	while ((dev = devs[i++]) != NULL) {
-		struct hid_device_info *tmp = hid_enumerate_from_libusb(dev, vendor_id, product_id);
+		struct hid_device_info *tmp = hid_enumerate_from_libusb(dev, vendor_id, product_id, 0);
 		if (cur_dev) {
 			cur_dev->next = tmp;
 		}
@@ -1168,7 +1172,7 @@ static int hid_libusb_hotplug_callback(libusb_context *ctx, libusb_device *devic
 static void process_hotplug_event(struct hid_hotplug_queue* msg)
 {
 	if (msg->event == LIBUSB_HOTPLUG_EVENT_DEVICE_ARRIVED) {
-		struct hid_device_info* info = hid_enumerate_from_libusb(msg->device, 0, 0);
+		struct hid_device_info* info = hid_enumerate_from_libusb(msg->device, 0, 0, 1);
 		struct hid_device_info* info_cur = info;
 		while (info_cur) {
 			/* For each device, call all matching callbacks */

```